general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_KDTREE_HPP
+#define MFEM_KDTREE_HPP
+
+#include <vector>
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <cmath>
+
+#include "../linalg/vector.hpp"
+
+namespace mfem
+{
+
+namespace KDTreeNorms
+{
+
+/// Evaluates l1 norm of a vector.
+template <typename Tfloat, int ndim>
+struct Norm_l1
+{
+   Tfloat tm;
+   Tfloat operator() (const Tfloat* xx)
+   {
+      if (xx[0]<Tfloat(0.0)) { tm=-xx[0];}
+      else { tm=xx[0];}
+      for (int i=1; i<ndim; i++)
+      {
+         if (xx[i]<Tfloat(0.0)) { tm=tm-xx[i]; }
+         else { tm=tm+xx[i];}
```

Comment:
abs might not be defined for the float data type. I was envisioning AD datatypes - However, if we limit the implementation to double only it can be replaced.

Reply:
Couldn't you add a definition of `abs` for any AD types that were missing it? This implementation has the same issues from an AD perspective about differentiability at zero. With `abs` defined, this whole function can be done with:
```c++
Tfloat tm(0);
for(int i =0; i < ndim; ++i) { tm += abs(xx[i]); }
return tm;
```
Why do you treat iteration zero differently?

Reply:
corrected.